
# This script combines multiple report files generated by the Bismark aligner
# and combines them into one single report files for further procesing. 
#

required_libraries <- c("foreach","optparse","stringr")
sapply(required_libraries, library, character.only=TRUE)

options(DelayedArray.block.size=4500e6)



option_list = list(
  make_option(c("-i","--input_report_files"),type='character',default=NULL,
              help="Output from bismark_methylation_extractor. Use comma-separated file names"),
  make_option(c("-s","--sample_name"),type='character',default=".",
              help="sample name for the files")
)
opt_parse=OptionParser(option_list=option_list)
opt = parse_args(opt_parse);

report_files = scan(opt$input_report_files, what="character")

extract <- function(pattern, lines) {
  idx <- grep(pattern, lines)[1]
  line <- lines[idx]
  m <- str_match(line, pattern) # The extracted match is in column 2
  if (ncol(m)==2) {
    return(m[1,2])
  } else {
    return("")
  }
}
patterns <- c(  total_reads = "Sequence pairs analysed in total:\t(.*)",
                uniquely_aligned_reads = "Number of paired-end alignments with a unique best hit:\t(.*)",
                non_uniquely_aligned_reads = "Sequence pairs did not map uniquely:\t(.*)",
                OT_reads = "CT/GA/CT:\t(.*)\t",
                OB_reads = "CT/GA/GA:\t(.*)\t",
                CTOT_reads = "GA/CT/CT:\t(.*)\t",
                CTOB_reads = "GA/CT/GA:\t(.*)\t",
                aligner_version = "\\(version: (.*)\\)",
                aligner_options = "specified options: (.*)",
                genome = "bisulfite genome of (.*) with",
                CHG_meth = "Total methylated C's in CHG context:\t(.*)",
                CHH_meth = "Total methylated C's in CHH context:\t(.*)",
                CHG_unmeth = "Total unmethylated C's in CHG context:\t(.*)",
                CHH_unmeth = "Total unmethylated C's in CHH context:\t(.*)"
)


getPhenoData <- function(pe_report_file) {
  bismark_report <- file.path(pe_report_file)
  samplename <- basename(bismark_report)
  samplename <- gsub(samplename, pattern="_PE_report.txt", replacement="")
  report_lines <- readLines(bismark_report)
  pd <- data.frame(row.names=samplename,
                   t(as.data.frame(sapply(patterns, extract, report_lines), stringsAsFactors=FALSE)), stringsAsFactors = FALSE)
  # Convert columns with only numeric values from string to numeric
  suppressWarnings(convert_idx <- which(!is.na(sapply(colnames(pd), function(colnam) as.numeric(pd[,colnam])))))
  for (idx in convert_idx) pd[,idx] <- as.numeric(pd[,idx])
  return(pd)
}

pd <- foreach(pe_report_file=report_files, .combine="rbind") %do% getPhenoData(pe_report_file)


filename = paste0(opt$sample_name,"_report.txt")
printer = file(filename,"w") 
write(sprintf("version: %s",pd$aligner_version[1]),printer,append=T)
write(sprintf("specified options: %s",pd$aligner_options[1]),printer,append=T)
write(sprintf("bisulfite genome of %s with",pd$genome[1]),printer,append=T)
write(sprintf("Sequence pairs analysed in total:\t %.f",sum(pd$total_reads)),printer,append=T)
write(sprintf("Number of paired-end alignments with a unique best hit:\t %.f",sum(pd$uniquely_aligned_reads)),printer,append=T)
write(sprintf("Sequence pairs did not map uniquely:\t %.f",sum(pd$non_uniquely_aligned_reads)),printer,append=T)
write(sprintf("CT/GA/CT:\t %.f",sum(pd$OT_reads)),printer,append=T)
write(sprintf("CT/GA/GA:\t %.f",sum(pd$OB_reads)),printer,append=T)
write(sprintf("GA/CT/CT:\t %.f",sum(pd$CTOT_reads)),printer,append=T)
write(sprintf("GA/CT/GA:\t %.f",sum(pd$CTOB_reads)),printer,append=T)
write(sprintf("Total methylated C's in CHG context:\t %.f",sum(pd$CHG_meth)),printer,append=T)
write(sprintf("Total methylated C's in CHH context:\t %.f",sum(pd$CHH_meth)),printer,append=T)
write(sprintf("Total unmethylated C's in CHG context:\t %.f",sum(pd$CHG_unmeth)),printer,append=T)
write(sprintf("Total unmethylated C's in CHH context:\t %.f",sum(pd$CHH_unmeth)),printer,append=T)
close(printer) 
